### PR TITLE
IODA converters feature observation redating and error amplification

### LINKED ISSUE #871: Time dependent obs error

### PR DESCRIPTION
**Pull request by apchoiCMD**
#### This PR has two features that are applied to RADS ADT IODA converters ONLY
- Changing the obs. datetime if those do not fall in DA cycle window (i.e. 6 hrs)
- <del>Increasing obs. error, however, the method is not yet decided (i.e. find the slope and increase linear error)</del>
- Linearly increasing obs. error by time dependent (<del>Errors are adjusted based on 0.1 meter per 6 hours from initial 0.1 meter</del>)
- Configuration will provide a slope of linear error

- Resolves #871

### CODE REVIEW

**Review by guillaumevernieres — COMMENTED** [review]
That needs to be redesigned.
The burden of providing the date bounds should be inside of the provider implementation. All that "redate" should do is swap dates that are before and after the bounds and adjust the obs error.  
Start with the RADS, that might be the only one that we need to implement.

**Comment by apchoiCMD** [comment]
Unfortunately, this option is not applicable with an existing sample of Rads, I will add another Rads sub-sample for making sure this work.

**Comment by apchoiCMD** [comment]
> Unfortunately, this option is not applicable with an existing sample of Rads, I will add another Rads sub-sample for making sure this work.

Two cdl files are replaced for the test.

File: utils/obsproc/NetCDFToIodaConverter.h [hunk 1]
```diff
@@ -255,6 +255,7 @@ namespace gdasapp {
     util::DateTime windowEnd_;
     std::vector<std::string> inputFilenames_;
     std::string outputFilename_;
+    std::string provider_;
```
Comment (guillaumevernieres) [inline utils/obsproc/NetCDFToIodaConverter.h]: not used, remove

File: utils/obsproc/Rads2Ioda.h [hunk 2]
```diff
@@ -28,6 +29,10 @@ namespace gdasapp {
     gdasapp::obsproc::iodavars::IodaVars providerToIodaVars(const std::string fileName) final {
       oops::Log::info() << "Processing files provided by the RADS" << std::endl;
 
+      // Get the window cycle from the configuration
```
Comment (guillaumevernieres) [inline utils/obsproc/Rads2Ioda.h]: why are you adding a new configuration? You have "window begin" and "window end" to work with.

File: utils/obsproc/Rads2Ioda.h [hunk 3]
```diff
@@ -123,7 +128,30 @@ namespace gdasapp {
...
+      int daysSinceEpoch = julianDate - 2400002;  // Since Epoch 1858-11-17
+      int secondsOffset = util::datefunctions::hmsToSeconds(hour, minute, second);
+
+      if (iodaVars.datetime_.size() == 0) {
+        oops::Log::info() << "datetime_ is empty" << std::endl;
+      } else {
+        // DA window begins
+        int64_t minDAwindow = (daysSinceEpoch*86400) + secondsOffset;
+        // DA window ends at hours of windowcyc later
+        int64_t maxDAwindow = minDAwindow + (windowcyc * 3600);
```
Comment (guillaumevernieres) [inline utils/obsproc/Rads2Ioda.h]: compute the "maxdawindow" the same way you computed "mindawindow". Add a utility function somewhere to de-clutter your code.

File: utils/obsproc/Rads2Ioda.h [hunk 4]
```diff
@@ -123,7 +128,30 @@ namespace gdasapp {
...
+
+      std::istringstream dabegin(windowBegin_.toString());
+      dabegin >> year >> dash >> month >> dash >> day >> T
+         >> hour >> colon >> minute >> colon >> second >> Z;
+
+      uint64_t julianDate = util::datefunctions::dateToJulian(year, month, day);
+      int daysSinceEpoch = julianDate - 2400002;  // Since Epoch 1858-11-17
+      int secondsOffset = util::datefunctions::hmsToSeconds(hour, minute, second);
+
+      if (iodaVars.datetime_.size() == 0) {
```
Comment (guillaumevernieres) [inline utils/obsproc/Rads2Ioda.h:131]: weird place for checking this ... Shouldn't that be earlier in the code?

File: utils/obsproc/util.h [hunk 5]
```diff
@@ -218,6 +218,24 @@ namespace gdasapp {
...
         }
+
+       // Swap the date
+       void reDate(int64_t minDAwindow, int64_t maxDAwindow, int windowcyc) {
+         for (int i = 0; i < location_; i++) {
+           if (datetime_(i) < minDAwindow) {
+             int timelatency = minDAwindow - datetime_(i);
+             datetime_(i) = minDAwindow + 1;  // for safety 1 sec
+             obsError_(i) += 0.1 / (3600 * windowcyc) * timelatency;
+           } else if (maxDAwindow < datetime_(i)) {
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: no need for an `else if` ... just add an other `if` statement.

File: utils/obsproc/util.h [hunk 6]
```diff
@@ -218,6 +218,24 @@ namespace gdasapp {
           oops::Log::test() << checksum(latitude_, "latitude") << std::endl;
           oops::Log::test() << checksum(datetime_, "datetime") << std::endl;
         }
+
+       // Swap the date
+       void reDate(int64_t minDAwindow, int64_t maxDAwindow, int windowcyc) {
+         for (int i = 0; i < location_; i++) {
+           if (datetime_(i) < minDAwindow) {
+             int timelatency = minDAwindow - datetime_(i);
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: "timelatency" does not describe what that variable is ... Keep it simple! dt deltat deltatime, ... so your peers understand what it is you are trying to.

File: utils/obsproc/util.h [hunk 7]
```diff
@@ -218,6 +218,24 @@ namespace gdasapp {
...
+             datetime_(i) = minDAwindow + 1;  // for safety 1 sec
+             obsError_(i) += 0.1 / (3600 * windowcyc) * timelatency;
+           } else if (maxDAwindow < datetime_(i)) {
+             int timelatency = datetime_(i) - maxDAwindow;
+             datetime_(i) = maxDAwindow - 1;  // for safety 1 sec
+             obsError_(i) += 0.1 / (3600 * windowcyc) * timelatency;
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: linear is fine, but pass the slope, in a meaningful way, through the configuration.

File: utils/test/testinput/gdas_rads2ioda.yaml [hunk 8]
```diff
@@ -1,6 +1,7 @@
 provider: RADS
-window begin: 2018-04-15T06:00:00Z
-window end: 2018-04-15T12:00:00Z
+window begin: 2021-06-30T21:00:00Z
+window end: 2021-07-01T03:00:00Z
+window cycle: 6
```
Comment (guillaumevernieres) [inline utils/test/testinput/gdas_rads2ioda.yaml]: remove "window cycle" from the config

**Review by guillaumevernieres — CHANGES_REQUESTED** [review]
See comments.

**Comment by AndrewEichmann-NOAA** [comment]
I have no specific suggestions, but it would probably be good to add a few comments about why some of this is being done

File: utils/obsproc/NetCDFToIodaConverter.h (hunk 1 above)
Comment (apchoiCMD) [reply utils/obsproc/NetCDFToIodaConverter.h]: > not used, remove

Removed.

File: utils/obsproc/Rads2Ioda.h (hunk 3 above)
Comment (apchoiCMD) [reply utils/obsproc/Rads2Ioda.h]: > compute the "maxdawindow" the same way you computed "mindawindow". Add a utility function somewhere to de-clutter your code.

Julian day conversion and sec is now placed in util.h

File: utils/obsproc/util.h (hunk 5 above)
Comment (apchoiCMD) [reply utils/obsproc/util.h]: Done.

File: utils/obsproc/util.h (hunk 6 above)
Comment (apchoiCMD) [reply utils/obsproc/util.h]: Done.

File: utils/obsproc/util.h (hunk 7 above)
Comment (apchoiCMD) [reply utils/obsproc/util.h]: declare initial error in configuration and its duration... slope will be defined by them.

File: utils/test/testinput/gdas_rads2ioda.yaml (hunk 8 above)
Comment (apchoiCMD) [reply utils/test/testinput/gdas_rads2ioda.yaml]: Done

File: utils/test/testinput/gdas_rads2ioda.yaml [hunk 9]
```diff
@@ -1,7 +1,9 @@
 provider: RADS
 window begin: 2021-06-30T21:00:00Z
 window end: 2021-07-01T03:00:00Z
-window cycle: 6
+error:
+  initial: 0.1
```
Comment (guillaumevernieres) [inline utils/test/testinput/gdas_rads2ioda.yaml]: no, the initial error should be the original one.

File: utils/test/testinput/gdas_rads2ioda.yaml [hunk 10]
```diff
@@ -1,7 +1,9 @@
...
+  duration of hours: 6
```
Comment (guillaumevernieres) [inline utils/test/testinput/gdas_rads2ioda.yaml]: Not very intuitive. How about having the slope as argument?

File: utils/obsproc/util.h [hunk 11]
```diff
@@ -219,19 +221,33 @@ namespace gdasapp {
           oops::Log::test() << checksum(datetime_, "datetime") << std::endl;
         }
 
-       // Swap the date
-       void reDate(int64_t minDAwindow, int64_t maxDAwindow, int windowcyc) {
+       // Convert JulianDate and second
+       void JulianDateTime(const std::string & str, int64_t & DAwindow) {
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: fine for now, but this should be a general utility if that doesn't already exist somewhere else.

File: utils/obsproc/util.h [hunk 12]
```diff
@@ -219,19 +221,33 @@ namespace gdasapp {
           oops::Log::test() << checksum(datetime_, "datetime") << std::endl;
         }
 
-       // Swap the date
-       void reDate(int64_t minDAwindow, int64_t maxDAwindow, int windowcyc) {
+       // Convert JulianDate and second
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: No sure I understand ...

File: utils/obsproc/util.h [hunk 13]
```diff
@@ -219,19 +221,33 @@ namespace gdasapp {
...
+       void JulianDateTime(const std::string & str, int64_t & DAwindow) {
+         int year, month, day, hour, minute, second;
+         util::datefunctions::stringToYYYYMMDDhhmmss(str, year, month, day, hour, minute, second);
+         uint64_t julianDate = util::datefunctions::dateToJulian(year, month, day);
+         // Since Epoch 1858-11-17
+         int daysSinceEpoch = julianDate - 2400002;
+         int secondsOffset = util::datefunctions::hmsToSeconds(hour, minute, second);
+         DAwindow = (daysSinceEpoch * 86400) + secondsOffset;
+       }
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: Can you explain clearly what `DAwindow` is here?

File: utils/obsproc/util.h [hunk 14]
```diff
@@ -219,19 +221,33 @@ namespace gdasapp {
...
+             datetime_(i) = minDAwindow + 1;
+             // one second is used for safety falling in min Da window
+             obsError_(i) += errInitial / (3600 * errDuration) * delta_t;
+           }
+           if (maxDAwindow < datetime_(i)) {
+             int delta_t = datetime_(i) - maxDAwindow;
+             datetime_(i) = maxDAwindow - 1;
+             // one second is used for safety falling in max Da window
+             obsError_(i) += errInitial / (3600 * errDuration) * delta_t;
+             // Errors are adjusted linearly based on 0.1 meter (initial) per 6 (duration) hours
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: seems convoluted ... how about just passing the rate of increase as argument instead?

File: utils/obsproc/Rads2Ioda.h (hunk 2 above)
Comment (AndrewEichmann-NOAA) [inline utils/obsproc/Rads2Ioda.h]: you probably want to delete this now

**Review by AndrewEichmann-NOAA — DISMISSED** [review]

File: utils/obsproc/util.h [hunk 15]
```diff
@@ -221,33 +221,32 @@ namespace gdasapp {
           oops::Log::test() << checksum(datetime_, "datetime") << std::endl;
         }
 
-       // Convert JulianDate and second
-       void JulianDateTime(const std::string & str, int64_t & DAwindow) {
+       // Convert ISO 8601 formatted datetime to epoch seconds
+       void DateToSeconds(const std::string & str, int64_t & start_end_seconds) {
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: the camel casing doesn't follow the proper convention, it should be `dateToSeconds` for a function. Also, sorry for being picky, but `start_end_seconds` for your variable name is confusing.

File: utils/obsproc/util.h [hunk 16]
```diff
@@ -218,6 +221,37 @@ namespace gdasapp {
...
+         int64_t minDAwindow = timeOffsets[0];
+         int64_t maxDAwindow = timeOffsets[1];
+         for (int i = 0; i < location_; i++) {
+           if (datetime_(i) < minDAwindow) {
+             int delta_t = minDAwindow - datetime_(i);
+             datetime_(i) = minDAwindow + 1;
+             // one second is used for safety falling in min DA window
+             obsError_(i) += errRatio * delta_t;
+           }
+           if (maxDAwindow < datetime_(i)) {
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h:245]: I had to look it up, I don't know why but putting the constant on the left through me off! Funnily enough, writing a conditional that way is called a `Yoda condition` ... Perfect for jedi work :)

**Review by guillaumevernieres — APPROVED** [review]
Looks good @apchoiCMD . We'll probably ask for "evidence" that the obs error are in fact adjusted the way we want, but seems OK from looking at the code.

**Review by AndrewEichmann-NOAA — APPROVED** [review]

File: utils/obsproc/util.h [hunk 17]
```diff
@@ -6,6 +6,9 @@
 #include <string>
 #include <vector>
 
+#include "ioda/../../../../core/IodaUtils.h"
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h:9]: @apchoiCMD , add ```find_package(ioda REQUIRED)``` in the cmakelists.txt and remove the relative path as @CoryMartin-NOAA pointed out last Thursday.

Comment (guillaumevernieres) [reply utils/obsproc/util.h:9]: I'm backtracking my review @apchoiCMD . The relative path you have is fine, unless you can find the header you need under `ioda/src/engines/ioda/include/ioda`.

**Review by guillaumevernieres — APPROVED** [review]
The `find_package` issue is not related to this PR, so clicking the green button!

**Pull request merged**